File: src/index_options.cpp
#include "index_options.h"

#include <absl/status/status.h>

#include <string_view>

#include "absl/status/statusor.h"
#include "absl/strings/numbers.h"
#include "absl/strings/str_format.h"
#include "re2/re2.h"

namespace vectorlite {

absl::StatusOr<IndexOptions> IndexOptions::FromString(
    std::string_view index_options) {
  static const re2::RE2 hnsw_reg("^hnsw\\((.*)\\)$");
  std::string key_value;
  if (!re2::RE2::FullMatch(index_options, hnsw_reg, &key_value)) {
    return absl::InvalidArgumentError(
        "Invalid index option. Only hnsw is supported");
  }

  IndexOptions options;
  static const re2::RE2 kv_reg("([\\w]+)=([\\w]+)");

  bool has_max_elements = false;
  std::string key;
  std::string value;

  std::string_view input(key_value);
  while (re2::RE2::FindAndConsume(&input, kv_reg, &key, &value)) {
    if (key == "max_elements") {
      if (!absl::SimpleAtoi<size_t>(value, &options.max_elements)) {
        std::string error =
            absl::StrFormat("Cannot parse max_elements: %s", value);
        return absl::InvalidArgumentError(error);
      }
      has_max_elements = true;
    } else if (key == "M") {
      if (!absl::SimpleAtoi<size_t>(value, &options.M)) {
        std::string error = absl::StrFormat("Cannot parse M: %s", value);
        return absl::InvalidArgumentError(error);
      }
    } else if (key == "ef_construction") {
      if (!absl::SimpleAtoi<size_t>(value, &options.ef_construction)) {
        std::string error =
            absl::StrFormat("Cannot parse ef_construction: %s", value);
        return absl::InvalidArgumentError(error);
      }
    } else if (key == "random_seed") {
      if (!absl::SimpleAtoi<size_t>(value, &options.random_seed)) {
        std::string error =
            absl::StrFormat("Cannot parse random_seed: %s", value);
        return absl::InvalidArgumentError(error);
      }
    } else if (key == "allow_replace_deleted") {
      if (!absl::SimpleAtob(value, &options.allow_replace_deleted)) {
        std::string error =
            absl::StrFormat("Cannot parse allow_replace_deleted: %s", value);
        return absl::InvalidArgumentError(error);
      }
    } else {
      std::string error = absl::StrFormat("Invalid index option: %s", key);
      return absl::InvalidArgumentError(error);
    }
  }

  if (!has_max_elements) {
    return absl::InvalidArgumentError(
        "max_elements is required but not provided");
  }
  return options;
}

}  // namespace vectorlite

File: rust/src/index_options.rs
//! Parsing of the HNSW index-options string, e.g.
//! `hnsw(max_elements=1000, M=16)`. Mirrors `index_options.cpp`.

#[derive(Clone, Debug)]
pub struct IndexOptions {
    pub max_elements: usize,
    pub m: usize,
    pub ef_construction: usize,
    pub random_seed: usize,
    pub allow_replace_deleted: bool,
}

impl Default for IndexOptions {
    fn default() -> Self {
        IndexOptions {
            max_elements: 0,
            m: 16,
            ef_construction: 200,
            random_seed: 100,
            allow_replace_deleted: true,
        }
    }
}

fn is_word(s: &str) -> bool {
    !s.is_empty() && s.chars().all(|c| c.is_ascii_alphanumeric() || c == '_')
}

/// Mirrors `absl::SimpleAtob`: accepts a small set of boolean spellings,
/// case-insensitively.
fn parse_bool(s: &str) -> Option<bool> {
    match s.to_ascii_lowercase().as_str() {
        "true" | "t" | "yes" | "y" | "on" | "1" => Some(true),
        "false" | "f" | "no" | "n" | "off" | "0" => Some(false),
        _ => None,
    }
}

impl IndexOptions {
    pub fn parse(input: &str) -> Result<IndexOptions, String> {
        let trimmed = input.trim();
        let inner = trimmed
            .strip_prefix("hnsw(")
            .and_then(|s| s.strip_suffix(')'))
            .ok_or_else(|| "Invalid index option. Only hnsw is supported".to_string())?;

        let mut options = IndexOptions::default();
        let mut has_max_elements = false;

        let body = inner.trim();
        if !body.is_empty() {
            for pair in body.split(',') {
                let pair = pair.trim();
                let (key, value) = pair.split_once('=').ok_or_else(|| {
                    format!(
                        "Invalid index option. Expected comma-separated key=value pairs, got: {inner}"
                    )
                })?;
                let key = key.trim();
                let value = value.trim();
                if !is_word(key) || !is_word(value) {
                    return Err(format!(
                        "Invalid index option. Expected comma-separated key=value pairs, got: {inner}"
                    ));
                }
                match key {
                    "max_elements" => {
                        options.max_elements = value
                            .parse()
                            .map_err(|_| format!("Cannot parse max_elements: {value}"))?;
                        has_max_elements = true;
                    }
                    "M" => {
                        options.m = value
                            .parse()
                            .map_err(|_| format!("Cannot parse M: {value}"))?;
                    }
                    "ef_construction" => {
                        options.ef_construction = value
                            .parse()
                            .map_err(|_| format!("Cannot parse ef_construction: {value}"))?;
                    }
                    "random_seed" => {
                        options.random_seed = value
                            .parse()
                            .map_err(|_| format!("Cannot parse random_seed: {value}"))?;
                    }
                    "allow_replace_deleted" => {
                        options.allow_replace_deleted = parse_bool(value).ok_or_else(|| {
                            format!("Cannot parse allow_replace_deleted: {value}")
                        })?;
                    }
                    other => {
                        return Err(format!("Invalid index option: {other}"));
                    }
                }
            }
        }

        if !has_max_elements {
            return Err("max_elements is required but not provided".to_string());
        }
        Ok(options)
    }
}
